src/crypto-providers/ledgerCryptoProvider.ts

```diff
@@ -580,12 +602,23 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
     const ttl = prepareTtl(body.ttl)
     const validityIntervalStart = prepareValidityIntervalStart(body.validityIntervalStart)
     const withdrawals = body.withdrawals?.map(
-      (withdrawal) => prepareWithdrawal(withdrawal, stakeSigningFiles),
+      (withdrawal) => prepareWithdrawal(withdrawal, stakeSigningFiles, signingMode),
     )
     const auxiliaryData = prepareMetaDataHashHex(body.metadataHash)
-
-    // Ledger expects to receive either a `null` mint field, or a non-empty array
-    const mint = (body.mint && body.mint.length > 0) ? prepareTokenBundle(body.mint) : null
+    const mint = body.mint ? prepareTokenBundle(body.mint) : null
+    const scriptDataHashHex = prepareScriptDataHash(body.scriptDataHash)
+    const collaterals = body.collaterals?.map(
+      (collateralInput, i) => prepareCollateralInput(
+        collateralInput, getSigningPath(paymentSigningFiles, inputs.length + i),
```

Comment:
This deserves an explanation in a comment. Why is `inputs` used here at all when we are dealing with collateral?

Reply:
The role of the function `getSigningPath` is unclear to me. Can we get rid of it, or refactor/rename/replace it somehow in a way that will make obvious what's going on?

Reply:
Comments added in https://github.com/vacuumlabs/cardano-hw-cli/pull/112/commits/39dca2c9884367933e86f6f1109495f4302ad9f0. I couldn't think of a simple and clear refactor, but I believe that the explanation in `getSigningPath` is pretty clear.

Reply:
If I understand correctly what is going on, it all seems very Potemkin-like... and indicates some potential problems with the APIs.

As far as I can see, ledgerjs does not require having a witness path attached to every input. (I've not checked about Trezor, but since multisig there are inputs possibly signed by several paths anyway, so I don't think there should be a requirement of having one of them directly attached to the input.)

Hence a more natural way for cardano-hw-cli would be to just pass all payment signing file paths in `additionalWitnessPaths`, even those for inputs. It truly does not know which path belongs to each input, so I guess it should not be trying to fake that knowledge.

Reply:
As discussed offline, we will resolve this separately. I created an issue for this: https://github.com/vacuumlabs/cardano-hw-cli/issues/120